enterprise/dev/ci/internal/ci/changed/diff.go

```diff
@@ -0,0 +1,137 @@
+package changed
+
+import (
+	"strings"
+)
+
+type Diff uint32
+
+const (
+	// None indicates no diff. Use sparingly.
+	None Diff = 0
+
+	Go Diff = 1 << iota
+	Client
+	GraphQL
+	DatabaseSchema
+	Docs
+	Dockerfiles
+	ExecutorDockerRegistryMirror
+	CIScripts
+	Terraform
+	SVG
+
+	// All indicates all changes should be considered included in this diff, except None.
+	All
+)
+
+func ParseDiff(files []string) (diff Diff) {
+	for _, p := range files {
+		// Affects Go
+		if strings.HasSuffix(p, ".go") || p == "go.sum" || p == "go.mod" {
+			diff |= Go
+		}
+
+		// Client
+		if !strings.HasSuffix(p, ".md") && (isRootClientFile(p) || strings.HasPrefix(p, "client/")) {
+			diff |= Client
+		}
+
+		// Affects GraphQL
+		if strings.HasSuffix(p, ".graphql") {
+			diff |= GraphQL
+		}
+
+		// Affects DB schema
+		if strings.HasPrefix(p, "migrations/") {
+			diff |= (DatabaseSchema | Go)
+		}
+
+		// Affects docs
+		if strings.HasPrefix(p, "doc/") && p != "CHANGELOG.md" {
+			diff |= Docs
+		}
+
+		// Affects Dockerfiles
+		if strings.HasPrefix(p, "Dockerfile") || strings.HasSuffix(p, "Dockerfile") {
+			diff |= Dockerfiles
+		}
+
+		// Affects executor docker registry mirror
+		if strings.HasPrefix(p, "enterprise/cmd/executor/docker-mirror/") {
+			diff |= ExecutorDockerRegistryMirror
+		}
+
+		// Affects CI scripts
+		if strings.HasPrefix(p, "enterprise/dev/ci/scripts") {
+			diff |= CIScripts
+		}
+
+		// Affects Terraform
+		if strings.HasSuffix(p, ".tf") {
+			diff |= Terraform
+		}
+
+		// Affects SVG files
+		if strings.HasSuffix(p, ".svg") {
+			diff |= SVG
+		}
+	}
+	return
+}
+
+func (d Diff) String() string {
+	switch d {
+	case None:
+		return "None"
+
+	case Go:
+		return "Go"
+	case Client:
+		return "Client"
+	case GraphQL:
+		return "GraphQL"
+	case DatabaseSchema:
+		return "DatabaseSchema"
+	case Docs:
+		return "Docs"
+	case Dockerfiles:
+		return "Dockerfiles"
+	case ExecutorDockerRegistryMirror:
+		return "ExecutorDockerRegistryMirror"
+	case CIScripts:
+		return "CIScripts"
+	case Terraform:
+		return "Terraform"
+	case SVG:
+		return "SVG"
+
+	case All:
+		return "All"
+	}
+
+	var allDiffs []string
+	for checkDiff := Go; checkDiff <= All<<1; checkDiff = 1 << checkDiff {
+		diffName := checkDiff.String()
+		if diffName != "" && d.Has(checkDiff) {
+			allDiffs = append(allDiffs, diffName)
+		}
+	}
+	return strings.Join(allDiffs, ", ")
+}
+
+// Has returns true if this diff includes the given diff.
```

Comment:
Slightly misleading. It returns true if any of the bits set in target are in d. If you want to check if target is a subset of d, then your check should be `d&target == target`. Given how you use it below you probably want to keep the current implementation, just use a better docstring.

Nit: rather than the word "this" and "given", use the words `d` and `target` in the docstring.

Reply:
```suggestion
// Has returns true if d has the target diff.
```

Does this sound more correct? I'm struggling to think of a better verb than `has` to describe this, definitely a bit rusty and a quick search  😅  or:

```suggestion
// Has returns true if the target diff corresponds to d.
```

Reply:
07bc74e3919c5843a0b23d53e7d17c72ac4774d1 for now